File: backend/app/api/screener.py
import json
from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Any, List, Optional

from app.services.screener import get_screener_service, list_presets, PRESETS

router = APIRouter()


class ScreenerRunRequest(BaseModel):
    preset_id: Optional[str] = None
    preset_ids: Optional[List[str]] = None
    conditions: Optional[List[dict[str, Any]]] = None
    limit: int = 50
    max_scan: Optional[int] = None
    include_chinext: bool = True
    include_star: bool = True
    include_bse: bool = True
    exclude_st: bool = True
    use_custom_pool: bool = False
    prefer_local_cache: bool = True


@router.get("/presets")
async def get_presets():
    return {"presets": list_presets()}


def _validate_screener_request(body: ScreenerRunRequest) -> None:
    ids = body.preset_ids or ([body.preset_id] if body.preset_id else [])
    if ids:
        unknown = [p for p in ids if p not in PRESETS]
        if unknown:
            raise HTTPException(status_code=400, detail=f"未知预设: {', '.join(unknown)}")
        return
    if not body.conditions:
        raise HTTPException(status_code=400, detail="请至少选择一个预设或提供自定义条件")


@router.post("/run")
async def run_screener(body: ScreenerRunRequest):
    _validate_screener_request(body)
    return get_screener_service().run_scan(
        preset_id=body.preset_id,
        preset_ids=body.preset_ids,
        conditions=body.conditions,
        limit=min(body.limit, 500),
        max_scan=body.max_scan if body.max_scan and body.max_scan > 0 else None,
        include_chinext=body.include_chinext,
        include_star=body.include_star,
        include_bse=body.include_bse,
        exclude_st=body.exclude_st,
        use_custom_pool=body.use_custom_pool,
        prefer_local_cache=body.prefer_local_cache,
    )


@router.get("/runs/{run_id}/results")
async def get_screener_run_results(
    run_id: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    try:
        return get_screener_service().get_run_results(run_id, page, page_size)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e


@router.post("/run/stream")
async def run_screener_stream(body: ScreenerRunRequest):
    _validate_screener_request(body)

    def event_stream():
        svc = get_screener_service()
        try:
            for item in svc.iter_scan(
                preset_id=body.preset_id,
                preset_ids=body.preset_ids,
                conditions=body.conditions,
                limit=min(body.limit, 500),
                max_scan=body.max_scan if body.max_scan and body.max_scan > 0 else None,
                include_chinext=body.include_chinext,
                include_star=body.include_star,
                include_bse=body.include_bse,
                exclude_st=body.exclude_st,
                use_custom_pool=body.use_custom_pool,
                prefer_local_cache=body.prefer_local_cache,
            ):
                yield f"data: {json.dumps(item, ensure_ascii=False, default=str)}\n\n"
        except Exception as e:
            err = {"event": "error", "error": str(e)}
            yield f"data: {json.dumps(err, ensure_ascii=False)}\n\n"

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )

File: frontend/src/api/screener.js
import { API_BASE_URL, createApiClient } from './http.js'
import { readSseStream } from './sse.js'

const api = createApiClient('/screener', { timeout: 300000 })

export default {
  presets() {
    return api.get('/presets').then((r) => r.data)
  },
  runResults(runId, page = 1, pageSize = 20) {
    return api
      .get(`/runs/${runId}/results`, { params: { page, page_size: pageSize } })
      .then((r) => r.data)
  },
  async runStream(presetIds, onEvent, options = {}) {
    const ids = Array.isArray(presetIds) ? presetIds : presetIds ? [presetIds] : []
    const response = await fetch(`${API_BASE_URL}/screener/run/stream`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Accept: 'text/event-stream' },
      body: JSON.stringify({
        preset_ids: ids.length ? ids : undefined,
        preset_id: ids.length === 1 ? ids[0] : undefined,
        ...(options.maxScan != null && options.maxScan > 0
          ? { max_scan: options.maxScan }
          : {}),
        include_chinext: options.includeChinext ?? false,
        include_star: options.includeStar ?? false,
        include_bse: options.includeBse ?? false,
        exclude_st: options.excludeSt ?? true,
        use_custom_pool: options.useCustomPool ?? false,
        prefer_local_cache: options.preferLocalCache ?? true
      })
    })
    if (!response.ok) throw new Error(await response.text())
    return readSseStream(response, {
      onEvent,
      isFinalEvent: (ev) => ev.event === 'complete'
    })
  }
}
